MennaMohie/assignment2/Solution.cpp

```diff
@@ -0,0 +1,174 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+template<class T>
+class BinaryTree;
+
+template<class T>
+class Node
+{
+    T value;
+    Node *left, *right;
+    friend class BinaryTree<T>;
+
+public:
+    Node():left(nullptr),right(nullptr) {}
+    Node(T val):left(nullptr),right(nullptr),value(val) {}
+
+    T getValue(){return value;}
+};
+
+template<class T>
+class BinaryTree
+{
+    Node<T> *root;
+
+public:
+    BinaryTree(T val):root(new Node<T>(val)) {}
+
+    Node<T>* getRoot()
+    {
+        return root;
+    }
+
+    bool contains(Node<T> *currentRoot, T val)
+    {
+        if(currentRoot == nullptr)
+            return false;
+        if(currentRoot->value == val)
+            return true;
+        return (contains(currentRoot->left, val) || contains(currentRoot->right, val));
+    }
+
+    bool resetRoot(T val)
+    {
+        if(root==nullptr)
+        {
+            root=new Node<T>(val);
+            return true;
+        }
+        else if(!contains(root, val))
+        {
+            root->value=val;
+            return true;
+        }
+        else
+            return false;
+    }
+
+    //adds a new node (val) as a left child to an existing node (parentVal)
+    bool addAsLeftChild(Node<T> *currentRoot, T parentVal, T val)
+    {
+        if(!contains(root, parentVal) || contains(root, val) || currentRoot==nullptr)
```

Comment:
Note that the function `contains` explores the whole tree, and you will do this on every step of the way.
It's better to split this function in 2, a public one that does the sanity checking once (and doesn't need the currentRoot parameter), and a private function that actually traverses the tree and does the insertion.
The same applies to the function leastCommonAncestor bellow.